README
NAME
    EV::ADNS - lightweight asynchronous dns queries using EV and libadns

SYNOPSIS
     use EV;
     use EV::ADNS;

     EV::ADNS::submit "example.com", EV::ADNS::r_addr, 0, sub {
        my ($status, $expires, @a) = @_;
        warn $a[0]; # "127.13.166.3" etc.
     };

     EV::loop;

DESCRIPTION
    This is a simple interface to libadns (asynchronous dns) that integrates
    well and automatically into the EV event loop. The documentation for
    libadns is vital to understand this module, see
    <http://www.chiark.greenend.org.uk/~ian/adns/>.

    You can use it only with EV (directly or indirectly, e.g. via Glib::EV).
    Apart from loading and using the "submit" function you need not do
    anything (except run an EV event loop).

OVERVIEW
    All the constants/enums from adns.h are available in the EV::ADNS
    namespace, without the "adns_" prefix, e.g. "adns_r_a" becomes
    "EV::ADNS::r_a", "adns__qtf_deref" becomes "EV::ADNS::_qtf_deref" and so
    on.

FUNCTIONS
    $query = EV::ADNS::submit "domain", $rrtype, $flags, $cb
        Submits a new request to be handled. See the "adns_submit" C
        function description for more details. The function optionally
        returns a query object which can be used to cancel an in-progress
        request. You do not need to store the query object, even if you
        ignore it the query will proceed.

        The callback will be invoked with a result status, the time the
        resource record validity expires and zero or more resource records,
        one scalar per result record. Example:

           sub adns_cb {
              my ($status, $expires, @rr) = @_;
              if ($status == EV::ADNS::s_ok) {
                 use JSON::XS;
                 warn encode_json \@rr;
              }
           }

        The format of result records varies considerably, here is some
        cursory documentation of how each record will look like, depending
        on the query type:

        EV::ADNS::r_a, EV::ADNS::r_addr
            An IPv4 address in dotted quad (string) form.

        EV::ADNS::r_ns_raw, EV::ADNS::r_cname, EV::ADNS::r_ptr,
        EV::ADNS::r_ptr_raw
            The resource record as a simple string.

        EV::ADNS::r_txt
            An arrayref of strings.

        EV::ADNS::r_ns
            A "host address", a hostname with any number of addresses (hint
            records).

            Currently only the hostname will be stored, so this is alway an
            arrayref with a single element of the hostname. Future versions
            might add additional address entries.

        EV::ADNS::r_hinfo
            An arrayref consisting of the two strings.

        EV::ADNS::r_rp, EV::ADNS::r_rp_raw
            An arrayref with two strings.

        EV::ADNS::r_mx
            An arrayref consisting of the priority and a "host address" (see
            "EV::ADNS::r_ns"). Example:

               [10, "mail10.example.com"]

        EV::ADNS::r_mx_raw
            An arrayref consisting of the priority and the hostname, e.g.
            "[10, "mail.example.com"]".

        EV::ADNS::r_soa, EV::ADNS::r_soa_raw
            An arrayref consisting of the primary nameserver, admin name,
            serial, refresh, retry expire and minimum times, e.g.:

              ["ns.example.net", "hostmaster@example.net", 2000001102, 86400, 21600, 2592000, 172800]

            The "raw" form doesn't mangle the e-mail address.

        EV::ADNS::r_srv_raw
            An arrayref consisting of the priority, weight, port and
            hostname, e.g.:

               [10, 10, 5060, "sip1.example.net"]

        EV::ADNS::r_srv
            The same as "EV::ADNS::r_srv_raw", but the hostname is replaced
            by a "host address" (see "EV::ADNS::r_ns").

        EV::ADNS::r_unknown
            A single octet string with the raw contents.

        anything else
            Currently "undef".

    $query->cancel
        Cancels a request that is in progress.

SEE ALSO
    EV, Net::ADNS another interface to adns, maybe better, but without real
    support to integrate it into other event loops.

AUTHOR
     Marc Lehmann <schmorp@schmorp.de>
     http://home.schmorp.de/



Changes
Revision history for EV::ADNS

2.2  Fri May  7 23:40:40 CEST 2010
        - avoid accumulating temporaries, which can "leak" memory
          in certain cases (testcase by Nikita Savin).

2.1  2008-05-17
	- support r_addr result type.

2.0  Sat Dec 22 17:49:53 CET 2007
	- upgrade to EV version 2.0 API.
        - add message about adns-1.4 being the minimum supported version.

1.0  Thu Dec 13 00:44:24 CET 2007
	- only run the prepare watcher if there actually is work to do
          (this also made the code simpler and faster).
	- possible future multiplicity support.

0.3  Mon Dec  3 20:02:40 CET 2007
	- properly call adns_processtimeouts in the timer callback.

0.2  Mon Dec  3 12:53:55 CET 2007
	- use an idle watcher to ensure that our prepare callback
          is called even when we submit requests late in the event loop.
        - doc fixes.

0.1  Sat Dec  1 23:35:27 CET 2007
	- original version, a Glib::EV clone.


Makefile.PL
use ExtUtils::MakeMaker;
use EV::MakeMaker;

print <<EOF;

***
*** This is an interface to the asynchronous dns resolver library, libadns
*** You need to have it installed before using this module.
***
*** This module has only been tested with adns-1.4, earlier versions might
*** work, but are not supported (upgrading is highly recommended due to
*** the many bugs in earlier versions).
***
*** For Debian GNU/Linux and Ubuntu, this would get you there:
*** apt-get install libadns1-dev
***

EOF

my $mm = MM->new({EV::MakeMaker::ev_args (
    dist         => {
       PREOP	=> 'pod2text ADNS.pm | tee README >$(DISTVNAME)/README; chmod -R u=rwX,go=rX . ;',
       COMPRESS	=> 'gzip -9v',
       SUFFIX	=> '.gz',
	            },
    NAME         => "EV::ADNS",
    VERSION_FROM => "ADNS.pm",
    LIBS         => ["-ladns"],
    PREREQ_FATAL => 1,
    PREREQ_PM    => {
       EV                  => 2.0,
    },
    EXTRA_META => q{
configure_requires:
  EV: 2.0
    },

)});

$mm->flush;



ADNS.pm
=head1 NAME

EV::ADNS - lightweight asynchronous dns queries using EV and libadns

=head1 SYNOPSIS

 use EV;
 use EV::ADNS;

 EV::ADNS::submit "example.com", EV::ADNS::r_addr, 0, sub {
    my ($status, $expires, @a) = @_;
    warn $a[0]; # "127.13.166.3" etc.
 };

 EV::loop;

=head1 DESCRIPTION

This is a simple interface to libadns (asynchronous dns) that
integrates well and automatically into the EV event loop. The
documentation for libadns is vital to understand this module, see
L<http://www.chiark.greenend.org.uk/~ian/adns/>.

You can use it only with EV (directly or indirectly, e.g. via
L<Glib::EV>). Apart from loading and using the C<submit> function you need
not do anything (except run an EV event loop).

=head1 OVERVIEW

All the constants/enums from F<adns.h> are available in the EV::ADNS
namespace, without the C<adns_> prefix, e.g. C<adns_r_a> becomes
C<EV::ADNS::r_a>, C<adns__qtf_deref> becomes C<EV::ADNS::_qtf_deref> and
so on.

=head1 FUNCTIONS

=over 4

=item $query = EV::ADNS::submit "domain", $rrtype, $flags, $cb

Submits a new request to be handled. See the C<adns_submit> C function
description for more details. The function optionally returns a query
object which can be used to cancel an in-progress request. You do not need
to store the query object, even if you ignore it the query will proceed.

The callback will be invoked with a result status, the time the resource
record validity expires and zero or more resource records, one scalar per
result record. Example:

   sub adns_cb {
      my ($status, $expires, @rr) = @_;
      if ($status == EV::ADNS::s_ok) {
         use JSON::XS;
         warn encode_json \@rr;
      }
   }

The format of result records varies considerably, here is some cursory
documentation of how each record will look like, depending on the query
type:

=over 4

=item EV::ADNS::r_a, EV::ADNS::r_addr

An IPv4 address in dotted quad (string) form.

=item EV::ADNS::r_ns_raw, EV::ADNS::r_cname, EV::ADNS::r_ptr, EV::ADNS::r_ptr_raw

The resource record as a simple string.

=item EV::ADNS::r_txt

An arrayref of strings.

=item EV::ADNS::r_ns

A "host address", a hostname with any number of addresses (hint records).

Currently only the hostname will be stored, so this is alway an arrayref
with a single element of the hostname. Future versions might add
additional address entries.

=item EV::ADNS::r_hinfo

An arrayref consisting of the two strings.

=item EV::ADNS::r_rp, EV::ADNS::r_rp_raw

An arrayref with two strings.

=item EV::ADNS::r_mx

An arrayref consisting of the priority and a "host address" (see
C<EV::ADNS::r_ns>). Example:

   [10, "mail10.example.com"]

=item EV::ADNS::r_mx_raw

An arrayref consisting of the priority and the hostname, e.g. C<[10,
"mail.example.com"]>.

=item EV::ADNS::r_soa, EV::ADNS::r_soa_raw

An arrayref consisting of the primary nameserver, admin name, serial,
refresh, retry expire and minimum times, e.g.:

  ["ns.example.net", "hostmaster@example.net", 2000001102, 86400, 21600, 2592000, 172800]

The "raw" form doesn't mangle the e-mail address.

=item EV::ADNS::r_srv_raw

An arrayref consisting of the priority, weight, port and hostname, e.g.:

   [10, 10, 5060, "sip1.example.net"]

=item EV::ADNS::r_srv

The same as C<EV::ADNS::r_srv_raw>, but the hostname is replaced by a "host
address" (see C<EV::ADNS::r_ns>).

=item EV::ADNS::r_unknown

A single octet string with the raw contents.

=item anything else

Currently C<undef>.

=back

=item $query->cancel

Cancels a request that is in progress.

=back

=cut

package EV::ADNS;

use Carp ();
use EV ();

BEGIN {
   $VERSION = '2.2';

   require XSLoader;
   XSLoader::load (EV::ADNS, $VERSION);
}

=head1 SEE ALSO

L<EV>, L<Net::ADNS> another interface to adns, maybe better, but without
real support to integrate it into other event loops.

=head1 AUTHOR

 Marc Lehmann <schmorp@schmorp.de>
 http://home.schmorp.de/

=cut

1

